syntax = "proto2";

option java_multiple_files = true;
option java_package = "com.facebook.encapp.proto";
option java_outer_classname = "TestDefinition";

message Common {
  optional string id = 1;
  optional string description = 2;
  optional string operation = 3;
  optional string start = 4;
}

enum DataValueType {
    stringType = 0;
    intType = 1;
    floatType = 2;
    longType = 3;
}

enum PixFmt {
  yuv420p = 0;
  yvu420p = 1;
  nv12 = 2;
  nv21 = 3;
  rgba = 4;
  p010le = 54;
}

enum FilterProcess {
    // A binary filter is a filter that will be
    // provided, build for the specific platform
    // and have a filepath in the Filter message
    binary = 0;
    // This is a filter that will be run on the host using ffmpeg
    // Settings in parameter.
    ffmpeg = 1;
    java = 2;
}

message Parameter {
    optional string key = 1;
    optional DataValueType type = 2;
    optional string value = 3;
    // For runtime purposes
    optional int64 framenum = 4;
}

// Filter can explicitly specify the input and output pix_fmt
// and sizes. If not specified, the input and output pix_fmt
// is assumed to follow the input and the configuration respectively.
message Filter {
    optional FilterProcess filter_process = 1;
    optional string filepath = 2;
    // This is a name identifying a specific filter
    // if the code contains more than one
    optional string method = 3;
    optional PixFmt input_pix_fmt = 4;
    optional PixFmt output_pix_fmt = 5;
    optional string input_resolution = 6;
    // If the outputfilter resolution satisfies the encoder configured
    // resolution this filter will act as a scaler
    optional string output_resolution = 7;
    // Arbitrary settings for the filter
    repeated Parameter parameter = 8;
}

message ProfileLevel {
    optional int32 profile = 1;
    optional int32 level = 2;
}

message Input {
  optional string filepath = 1;
  optional string resolution = 2;
  optional PixFmt pix_fmt = 3;
  optional float framerate = 4;
  optional int32 playout_frames = 5;
  optional int32 pursuit = 6;
  optional bool realtime = 7;
  optional float stoptime_sec = 8;
  optional bool show = 9;
}

// Preprocess input
// Depending on FilterProcess it can happen on 
// host or on device
message PreProcess {
    optional Filter filter = 1;
}

message Configure {
  repeated Parameter parameter = 1;
  // known shortcuts
  optional string codec = 2;
  optional bool encode = 3;
  optional bool surface = 4;
  optional string mime = 5;
  optional string bitrate = 6;
  enum BitrateMode {
    cq = 0;
    vbr = 1;
    cbr = 2;
    cbr_fd = 3;
  }
  optional BitrateMode bitrate_mode = 7;
  optional uint64 durationUs = 8;
  optional string resolution = 9;

  optional int32 color_format = 10;
  enum ColorStandard {
    bt601_pal = 2;
    bt601_ntsc = 4;
    bt709 = 1;
    bt2020 = 6;
  }
  optional ColorStandard color_standard = 11;
  enum ColorRange {
    full = 1;
    limited = 2;
  }
  optional ColorRange color_range = 12;
  enum ColorTransfer {
    linear = 1;
    sdr_video = 3;
    st2084 = 6;
    hlg = 7;
  }
  optional ColorTransfer color_transfer = 13;
  optional string color_transfer_request = 14;
  optional float framerate = 15;
  optional int32 i_frame_interval = 16;
  optional int32 intra_refresh_period = 17;
  optional int32 latency = 18;
  optional int64 repeat_previous_frame_after = 19;
  optional string ts_schema = 20;
  optional int32 quality = 21;
  optional int32 complexity = 22;
  optional bool decode_dump = 23 [default = false];
  optional ProfileLevel profile_level = 24;
}

message Runtime {
  repeated Parameter parameter = 1;
  // known shortcuts
  message VideoBitrateParameter {
    optional int64 framenum = 1;
    //can be just a figure in bps or with prefix, k or M
    optional string bitrate = 2;
  }
  repeated VideoBitrateParameter video_bitrate = 2;
  repeated int64 drop = 3;
  message DynamicFramerateParameter {
    optional int64 framenum = 1;
    optional float framerate = 2;
  }
  repeated DynamicFramerateParameter dynamic_framerate = 4;
  repeated int64 request_sync = 5;
}

message DecoderConfigure {
  repeated Parameter parameter = 1;
  optional string codec = 2;
}

message DecoderRuntime {
  repeated Parameter parameter = 1;
}

message Parallel {
  repeated Test test = 1;
}

message Serial {
  repeated Test test = 1;
}

message Test {
  optional Common common = 1;
  optional Input input = 2;
  optional Configure configure = 3;
  optional Runtime runtime = 4;
  optional DecoderConfigure decoder_configure = 5;
  optional DecoderRuntime decoder_runtime = 6;
  optional Parallel parallel = 7;
  optional PreProcess preprocess = 8;
}

message TestSuite {
  // Tests are serial unless included in another test as parallel
  repeated Test test = 1;
}
